### ISSUE TITLE
Feature Request: OCR shopping receipts

### ISSUE THREAD

**Issue by Sohalt** [grocy/grocy#404]
It would be cool to have the option to scan a receipt using OCR to automatically add the listed products and prices.

**Comment by DeeeeLAN** [comment]
[Looks like there is an NPM package available](https://www.npmjs.com/package/ocr), although I don't know how much work it would be to grab the right information. Most of my receipts use somewhat cryptic codes for the items as well, which would be fine I guess if they are added as barcodes to the appropriate items.

**Comment by berrnd** [comment]
Thanks for the suggestion, but the mentioned package has requirements beyond what we have (a Browser and PHP)...

> node -v x64 8.0 or greater and npm.
> gcc --version 4.8.1 or greater.
> g++ -v 4.8.1 or greater.
>  java -version x64 JRE 1.8

**Comment by jrester** [comment]
What about [tesseract](https://github.com/tesseract-ocr/tesseract) with [tesseract-ocr-for-php](https://github.com/thiagoalessio/tesseract-ocr-for-php)? The binaries are available for many distros and could also be easily integrated into the docker image as well as the php package can be added via composer.

**Comment by dvankley** [comment]
I'm also curious about this. Many retailers have e-receipt systems, so an alternative approach might be plugins per-retailer to import e-receipts and lookup based on UPC?

**Comment by dvankley** [comment]
The issue with OCR'ing paper receipts (beyond the usual challenges of OCR) is that many of them don't have UPC codes on them, so you have the challenge of trying to resolve the product name based on whatever arbitrary abbreviation that retailer uses for that product. Each retailer also has their own receipt format, which further complicates things.

That said, I think a combination of scraping e-receipts and OCR'ing paper receipts that do have barcodes on them (i.e. Costco) might be feasible. I'll play around with this when I have some time.

**Comment by jrester** [comment]
But I am not sure if this would be a good addition to grocy, because it is quite heavy. Maybe it would be better to run this as a standalone service and just use the grocy API like the barcode scanner app does.

**Comment by dvankley** [comment]
> But I am not sure if this would be a good addition to grocy, because it is quite heavy. Maybe it would be better to run this as a standalone service and just use the grocy API like the barcode scanner app does.
> So those problems you mentioned could be solved in a different place rather than blowing up the codebase of grocy.

Yeah, you're probably right on that.

**Comment by thost96** [comment]
> 

could you explain how you used paperless for that?

**Comment by Pheggas** [comment]
> > Alright, thanx. I got an idea of connecting [paperless](https://github.com/the-paperless-project/paperless) and Grocy. Paperless has OCR so no need to create one directly in Grocy and also, it wouldn't be heavy when it has no OCR but instead, has just connected 3rd party app which OCR scan every receipt.
> 
> could you explain how you used paperless for that?

It would send that receipt to paperless for OCR and output PDF would go back to Grocy. I use paperless for all my documents and it works flawlessly. I think it wouldn't be a big problem.

**Comment by NorkzYT** [comment]
> > > Alright, thanx. I got an idea of connecting [paperless](https://github.com/the-paperless-project/paperless) and Grocy. Paperless has OCR so no need to create one directly in Grocy and also, it wouldn't be heavy when it has no OCR but instead, has just connected 3rd party app which OCR scan every receipt.
> > 
> > 
> > could you explain how you used paperless for that?
> 
> It would send that receipt to paperless for OCR and output PDF would go back to Grocy. I use paperless for all my documents and it works flawlessly. I think it wouldn't be a big problem.

Hello mate, have you given your idea a spin? I am quite interested in the entire process.

**Comment by Pheggas** [comment]
> Hello mate, have you given your idea a spin? I am quite interested in the entire process.

Not quite. It was a long time ago and I pretty much fell off that idea.

**Comment by NorkzYT** [comment]
> > Hello mate, have you given your idea a spin? I am quite interested in the entire process.
> 
> Not quite. It was a long time ago and I pretty much fell off that idea.

Understood.

**Comment by paulcs** [comment]
I actually am thinking of making something like this seperate from the main app using scanned and properly tagged receipts within Paperless-ngx's API to import directly into Grocy. See here for more info: goto [Reddit thread](https://www.reddit.com/r/grocy/comments/1nb5mh7/paperlessngx_receipts_automated_import_into_grocy/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).

**Comment by toxic0berliner** [comment]
This would be awesome. 
I know Google already does this very well in their reward app. 
Having commercial use I fear such a project in foss would be heavyly copied by commercial actors but maybe one might actually contribute...
